leakcanary-android-core/src/main/java/leakcanary/ConfigBuilder.kt

```diff
@@ -0,0 +1,90 @@
+package leakcanary
+
+import shark.LeakingObjectFinder
+import shark.MetadataExtractor
+import shark.ObjectInspector
+import shark.ReferenceMatcher
+
+/**
+ * Builder for [LeakCanary.Config] intended to use only by Java callers.
+ *
+ * Usage:
+ * ```
+ * LeakCanary.Config config = LeakCanary.getConfig().newBuilder()
+ *    .dumpHeap(false)
+ *    .retainedVisibleThreshold(3)
+ *    .maxStoredHeapDumps(10)
+ *    .build();
+ * LeakCanary.setConfig(config);
+ * ```
+ *
+ * Kotlin callers should be using `copy()` method instead:
+ * ```
+ * LeakCanary.config = LeakCanary.config.copy(dumpHeap = false)
+ * ```
+ */
+@Suppress("TooManyFunctions")
+class ConfigBuilder(config: LeakCanary.Config) {
+  private var dumpHeap: Boolean = config.dumpHeap
+  private var dumpHeapWhenDebugging: Boolean = config.dumpHeapWhenDebugging
+  private var retainedVisibleThreshold: Int = config.retainedVisibleThreshold
+  private var referenceMatchers: List<ReferenceMatcher> = config.referenceMatchers
+  private var objectInspectors: List<ObjectInspector> = config.objectInspectors
+  private var onHeapAnalyzedListener: OnHeapAnalyzedListener = config.onHeapAnalyzedListener
+  private var metadataExtractor: MetadataExtractor = config.metadataExtractor
+  private var computeRetainedHeapSize: Boolean = config.computeRetainedHeapSize
+  private var maxStoredHeapDumps: Int = config.maxStoredHeapDumps
+  private var requestWriteExternalStoragePermission: Boolean =
+    config.requestWriteExternalStoragePermission
+  private var leakingObjectFinder: LeakingObjectFinder = config.leakingObjectFinder
+
+  /** @see [LeakCanary.Config.dumpHeap] */
+  fun dumpHeap(value: Boolean) = apply { dumpHeap = value }
```

Comment:
Still not sure about the parameter name here and below. `value` is short and coupled with method name is more that enough, but still feels wrong to me..

Reply:
why not name it the same as the property, ie dumpHeap, then `this. dumpHeap = dumpHeap` ?

Reply:
It feels a bit explicit, but I guess that's a better code style after all.
Guess, I just don't like the builder pattern in general - too much boilerplate!